flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/shuffle/TieredInternalShuffleMaster.java

```diff
@@ -0,0 +1,85 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition.hybrid.tiered.shuffle;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.common.TieredStorageConfiguration;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage.ResourceRegistry;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage.TieredStorageMasterClient;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.tier.TierFactory;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.tier.TierMasterAgent;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.runtime.io.network.partition.hybrid.tiered.common.TieredStorageIdMappingUtils.convertId;
+
+/**
+ * A wrapper internal shuffle master class for tiered storage. All the tiered storage operations
+ * with the shuffle master should be wrapped in this class.
+ */
+public class TieredInternalShuffleMaster {
+
+    private final TieredStorageMasterClient tieredStorageMasterClient;
+
+    private final Map<JobID, List<ResultPartitionID>> jobPartitionIds;
+
+    public TieredInternalShuffleMaster(Configuration conf) {
+        TieredStorageConfiguration tieredStorageConfiguration =
+                createTieredStorageConfiguration(conf);
+        ResourceRegistry resourceRegistry = new ResourceRegistry();
+        List<TierMasterAgent> tierFactories =
+                tieredStorageConfiguration.getTierFactories().stream()
+                        .map(TierFactory::createMasterAgent)
+                        .peek(tierMasterAgent -> tierMasterAgent.setup(resourceRegistry))
+                        .collect(Collectors.toList());
+        this.tieredStorageMasterClient = new TieredStorageMasterClient(tierFactories);
+        this.jobPartitionIds = new HashMap<>();
+    }
+
+    public void addPartition(JobID jobID, ResultPartitionID resultPartitionID) {
+        jobPartitionIds.computeIfAbsent(jobID, ignore -> new ArrayList<>()).add(resultPartitionID);
+        tieredStorageMasterClient.addPartition(convertId(resultPartitionID));
+    }
+
+    public void releasePartition(ResultPartitionID resultPartitionID) {
+        tieredStorageMasterClient.releasePartition(convertId(resultPartitionID));
+    }
+
+    public void unregisterJob(JobID jobID) {
+        List<ResultPartitionID> resultPartitionIDs = jobPartitionIds.remove(jobID);
+        if (resultPartitionIDs != null) {
+            resultPartitionIDs.forEach(
+                    resultPartitionID ->
+                            tieredStorageMasterClient.releasePartition(
+                                    convertId(resultPartitionID)));
+        }
+    }
```

Comment:
Each partition can be released for multiple times, one from `releasePartition` and one from `unregisterJob`, let alone the possibility that `releasePartition` and `unregisterJob` may or may not be called multiple times for the same partition/job id. This means all the classes handling the release of partitions (`TieredStorageMasterClient`, `TierMasterAgent`, etc.) would need to be aware of such possibility.

Alternatively, we can guarantee each partition is only released for once, by eagerly removing the partition from `releasePartition ` in `releasePartition` and calling `TieredStorageMasterClient#releasePartition` only if the partition can be found in `releasePartition`.

Reply:
OK. Because we can not know the exact jobId when releasing a result partition, I added a `partitionJobIds` to record the jobId of the `resultPartitioinId` to be released.

1. When adding a partition, add the partition id and job id to `jobPartitionIds` and `partitionJobIds` separately.
2. When releasing a partition, remove the partition id in the `jobPartitionIds` and `partitionJobIds` separately.
3. When registering a job, remove all the partition ids of the job in the `jobPartitionIds` and `partitionJobIds`.